### PR TITLE
Fix Component Mode border flicker on changing component mode

### PR DESCRIPTION
**Pull request by amzn-ahmadkrm**
## What does this PR do?

_Please describe your PR. For a bug fix, what was the old behavior, what is the new behavior?_

When the user changes component mode via the switcher, the border flickered on the change, with this PR, the border no longer flickers.

Before:
https://user-images.githubusercontent.com/105638312/185807369-bb841254-eb94-4722-9d7f-f134c99814e0.mp4

After:
https://user-images.githubusercontent.com/105638312/185807390-3c4b0670-fc34-4632-8ba2-5925c6724bf3.mp4

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp:250]: suggestion: A guide comment here to indicate what transform mode is special would definitely be helpful for future maintainers

File: Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/EditorBaseComponentMode.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/EditorBaseComponentMode.cpp:57]: update: Please always initialize bus result variables otherwise we may have an uninitialized variable read if the bus isn't connected or isn't found. So this would be `bool borderVisible = false`.

File: Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/EditorComponentModeBus.h
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/EditorComponentModeBus.h:145]: update: In code it's fine to just call it Component Mode (not Component Edit Mode) to be more consistent

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp:318]: update: Again a guide comment here would be useful to explain _why_ we're doing this

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp:321]: It looks like this and the code just above should also just call `ChangeViewportBorderText` below

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp:464]: update: Again a guide comment here for why we're doing this would be super helpful otherwise it's hard to know why it's required and what it's going to do

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.h
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.h:97]: update: This should be be returning a `const bool`, the function should be marked `const`. So this should be `bool ViewportBorderVisible() bool` (it would also be good to move the implementation out of line to keep this interface cleaner)

```
inline bool ViewportUiDisplay::ViewportBorderVisible() const
{
    ...
}
```

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiManager.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiManager.cpp:269]: update: Change mentioned earlier about the const function

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiManager.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiManager.cpp:274]: update: We should use a `string_view` or `const std::string` here instead

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiRequestBus.h
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiRequestBus.h:107]: update: Better to use phrasing to say what this does, not what it's doing. So this would be `Returns if the viewport border is visible.`

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiRequestBus.h
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiRequestBus.h:111]: nit: This would be `Removes viewport border` if we can change it while we're here 🙂

**Review by hultonha — COMMENTED** [review]
Heading in the right direction just a few small things to fix. Thanks!

File: Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp
Comment (hultonha) [reply Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp:253]: way better now thanks! 🙂

File: Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/EditorComponentModeBus.h
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/EditorComponentModeBus.h:146]: suggestion: This section of code doesn't really know anything about the switcher, so referencing it here might not be a good idea. Just saying `Switches the Component Mode immediately` (as this should work from somewhere else that might not want to use the switcher in future)

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp:329]: Thanks for adding this comment 👍

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp:346]: update: This does not need to be marked `inline` in the .cpp (only if it's in the header) and you can remove the const from the the bool return true (`const bool` -> `bool`)

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp
Comment (hultonha) [reply Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiDisplay.cpp:471]: Great comment update thanks 👍

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiRequestBus.h
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiRequestBus.h:108]: update: `const bool` -> `bool`

**Review by hultonha — APPROVED** [review]
Change looks good, approving with just a couple small comments to address. Thanks!

**Comment by hultonha** [comment]
Great work looking into the tests and I'm sure we can get this working, but for now it is probably worth extracting them and moving to a separate PR and we can get this fix and improvement in sooner rather than later (we can catch-up first thing next week to look at the tests as a separate PR). Thanks @amzn-ahmadkrm 👍

**Comment by amzn-ahmadkrm** [comment]
Removing unit test to add in a follow up PR

**Review by jonawals — COMMENTED** [review]
LGTM

File: Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/EditorComponentModeBus.h
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/EditorComponentModeBus.h:145]: uber nit: Sorry this need three `/` to be a Doxygen comment (this is one of the reasons why we now prefer `//!` but we don't need to make that change in this file yet (could be a separate formatting style change later)

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiManager.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiManager.cpp:274]: update: Please can this be a `const AZStd::string&` (so we're not making a copy here)

**Review by hultonha — APPROVED** [review]
Two super minor comments but otherwise looks good :+1: Be great to do the tests as a fast follow after, thanks!

**Review by jonawals — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiRequestBus.h
Comment (greerdv) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ViewportUi/ViewportUiRequestBus.h:107]: Tiny nitpick - typo in "Returns"

**Review by greerdv — APPROVED** [review]
LGTM, nice work!

**Pull request merged**
